### ISSUE TITLE
Add option to generate C# nullable annotations

### ISSUE THREAD

**Issue by YairHalberstadt** [protocolbuffers/protobuf#6632]
**Language** : C#

**Feature Request**

C# 8.0 adds support for [nullability annotations](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references).

It would be helpful if the the generated protobuf objects added these annotations. This would provide warnings if people tried to dereference a possibly null value, or for example, tried to assign null to a string property of a protobuf object.

 At first sight it looks like the rules for which types can be nullable and which can't should be quite simple.

I think the first stage would be annotating the Protobuf Nuget Packages.

Of course nullable annotaions are a compile error in earlier versions of C# than c# 8.0. Therefore this has to be off by default, and triggered by a command line option.

It would be helpful if this was added for GRPC as well.

**Comment by jtattermusch** [comment]
I'm not generally opposed to having this, but here are the challenges:
1. we would need a clear design - feel free to provide more thoughts here if you want
2. currently this is not a priority so there's no ETA and implementation would need to be contributed by someone from the community (but we'd need a design for this feature first).
3. C# 8 nullable annotations are a brand new C# 8 feature, but we definitely need to maintain the backward compatibility - which probably also means that it's going to take a while before this can actually be used. 

There are two separate areas:
1. annotations in the Google.Protobuf  runtime  (=the nuget) - then concern is backward-compatibility of the APIs (you cannot have an "option" for enabling/disabling this so it would need to work for everyone - and it's going to take a very long time until everyone can use C# 8).
2. annotations in the generated code (you can have a codegen option to enable the default-off feature), the question is how useful these checks would be - it's generated code and most of it is just field accessors, so I'm not sure we'd gain much here (a more detailed design would help).

**Comment by YairHalberstadt** [comment]
> annotations in the Google.Protobuf runtime (=the nuget) - then concern is backward-compatibility of the APIs (you cannot have an "option" for enabling/disabling this so it would need to work for everyone - and it's going to take a very long time until everyone can use C# 8).

The annotations are backwards compatible, so I don't think this should be an issue.

> annotations in the generated code (you can have a codegen option to enable the default-off feature), the question is how useful these checks would be - it's generated code and most of it is just field accessors, so I'm not sure we'd gain much here (a more detailed design would help).

A very common mistake I make is to set a string to null. That would be the single most useful gain.

With GRPC specifically, I often by mistake return null from the server implementation. That would be another gain but of course that is for a different repository.

I don't know a huge amount about protoc but my initial design would be:

All string properties are marked non-null.
No change in struct properties.
All collection properties (repeated, or maps) are marked as non-null.
All other properties are marked as nullable.

Parser property marked as non-null.
Descriptor property marked as non-null.

Constructor parameter is non-null
Clone method returns non-null
Equals parameter is nullable
Mergefrom parameter is nullable
MergeFrom(CodedInputStream) parameter is non-null
WriteTo parameter is non-null

**Comment by ObsidianMinor** [comment]
Instead of using the normal C# 8 syntax perhaps we could make use of the [code analysis attributes](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis?view=netstandard-2.1), that way the syntax works no matter what version of C# you're using. Of course that'd require that a separate package (like [Microsoft.Bcl.Async](https://www.nuget.org/packages/Microsoft.Bcl.Async/)) for everything under .NET Standard 2.1 but should that ever happen it could probably be done that way.

**Comment by YairHalberstadt** [comment]
@ObsidianMinor

I think that would be highly undesirable. C# 8.0 introduced nullability as a first class feature, and the amount of work that's been put in is phenomenal (probably more than any other feature in the history of C#). Even then it leaves a large number of holes.

Existing nullability attributes are extremely limited in what they can express, and enforce, and would fail to integrate well with C# 8.0. I think that it's better to make nullability work properly for a subset of users than to make it work badly for a larger subset.

**Comment by ObsidianMinor** [comment]
Were those attributes not specifically added for C# 8? dotnet/corefx#37826 says otherwise. They work with C# 8 flawlessly.

```cs
public string TypeUrl {
  get { return typeUrl_; }
  [global::System.Diagnostics.CodeAnalysis.DisallowNullAttribute]
  set {
    typeUrl_ = pb::ProtoPreconditions.CheckNotNull(value, "value");
  }
}

// .. somewhere far far away
#nullable enable

any.TypeUrl = null; // Cannot convert null literal to non-nullable reference type
```

All of the cases you mentioned work with these attributes.

String properties => DisallowNullAttribute.
All collection properties (repeated, or maps) => NotNullAttribute.
All other properties => MaybeNull.

Parser property => NotNullAttribute.
Descriptor property => NotNullAttribute.

Constructor parameter => DisallowNullAttribute
Clone method  => NotNullAttribute
Equals parameter => AllowNullAttribute
MergeFrom parameter => AllowNullAttribute
MergeFrom(CodedInputStream)  => DisallowNullAttribute
WriteTo  => DisallowNullAttribute

**Comment by YairHalberstadt** [comment]
@ObsidianMinor

I see I misunderstood which attributes you were referring to. My apologies.

The C# 8 nullable annotations emit attributes themselves. The coreFX attributes are meant for when the behaviour is to complex to be described by these annotations.

In so far as pre C# 8 tooling understands the CoreFX attributes, I would expect it to understand the attributes emitted by the compiler. However I strongly doubt either would work well.

**Comment by ObsidianMinor** [comment]
I can probably get a prototype of this working pretty easily but it could only be pulled when and if dotnet/corefx#40947 happens.

Using these annotations would allow us to do this in a backwards compatible way. The behavior of nulls in generated code and the library is too complex to be done with standard `?` and `!` syntax. So rather than doing it using that syntax, we can express what we need using the attributes. This has the included bonus of working on other lang versions. Only downside is this requires the nuget package as shown above.

**Comment by YairHalberstadt** [comment]
> Using these annotations would allow us to do this in a backwards compatible way.

Ah you mean to avoid a switch? That sounds like a decent idea.

**Comment by ObsidianMinor** [comment]
Alright I wrote a prototype (untested) on my fork on the [csharp/nullability](https://github.com/ObsidianMinor/protobuf/tree/csharp/nullability) branch. As it turns out nullability attributes do not affect code analysis in function bodies as noted in dotnet/roslyn#36039 so depending on whether the attributes affect code analysis in the future we may or may not have to disable warnings on the file level which is a very simple addition. Again, this will only work if a separate nuget package is made for these attributes, it works on the prototype since I manually copy-pasted the attributes into Google.Protobuf.

**Comment by jtattermusch** [comment]
I don't think we'd want to add another dependency to be able to reference the attributes.

**Comment by LiamMorrow** [comment]
This seems like a fairly decent compromise in reference the the idea of not wishing to add another dependency:
https://www.nuget.org/packages/Nullable/
Essentially does what @ObsidianMinor did where the attributes get compiled into the application without having to copy them into the repo.  Unless the preference is to not have any new dependencies, regardless of if they then cause a dependency downstream to consuming projects.

**Comment by LiamMorrow** [comment]
This is an interesting problem since it would be a breaking change (not so much binary breaking as source breaking).  Consider a project with nullable enabled that has code like this:
```proto
message FooMessage {
  google.protobuf.StringValue bar = 1;
}
```
```csharp
void DoSomething(FooMessage foo) 
{
    string extractedBar = foo.Bar;
    // do something with that string
}
```
Previously that would have emitted no warnings, but with a change like this the compiler will now say that we are possibly assigning null to a non nullable (`extractedBar`).  Now clearly this is the desired behavior (this is the whole point of nullable reference types).  The issue is when we look at this scenario in a project which has `TreatWarningsAsErrrors` enabled.  When we update the tooling reference, our project will no longer compile.  This would have a knock on effect with libraries that export their protobuf types (thinking gax etc.).

I'm strongly in favor of having the generator output nullable annotations, however this seems like it could have larger consequences than meets the eye, and thought it would be worth mentioning for discussion since an official proposal has not been done up.

**Comment by YairHalberstadt** [comment]
@LiamMorrow 
That is true of every single library. The official line of the C# team is: nullable reference types will be disabled by default till .Net 5. If you enable them earlier, expect some churn till then whilst libraries switch NRTs on.

**Comment by LiamMorrow** [comment]
Very true @YairHalberstadt .  I should have put a note I was mentioning it more of a justification to others (and my future self) for why it will likely be a while before this is turned on by default.   I am still interested in the option of a flag to enable it for generation.  Seems like it could be a good stopgap in the mean time.

**Comment by Falco20019** [comment]
We are also interested in this feature. We currently look into enabling nullables and being able to differentiate between `string` and `google.protobuf.StringValue` would be a great benefit for us.

I fully understand that this will take some more time, especially since most people won't enable it until it has a better adoption in libs. Just wanted to give it an upvote.

**Comment by thefringeninja** [comment]
In the meantime, could we have the auto generated code insert `#nullable disable` at the top of each generated file?

**Comment by jtattermusch** [comment]
@thefringeninja  I think adding '#nullable disable' to the generated files would actually break compilation on older compilers (since they don't know what `#nullable` means) and that's completely unacceptable (unless you can somehow include `#nullable` conditionally).

**Comment by thefringeninja** [comment]
I'm not sure about .net framework, but all the dotnet core versions that don't support `#nullable` are already EOL.

Even so, a flag to enable this would be more than sufficient.

**Comment by jtattermusch** [comment]
> I'm not sure about .net framework, but all the dotnet core versions that don't support `#nullable` are already EOL.

Not true, .NET Core 2.1 is supported until August 2021 and it doesn't support nullable reference types.
Even if it did, Google.Protobuf supports everything starting from net45 and we cannot just break the generated code for existing users.
> 
> Even so, a flag to enable this would be more than sufficient.

New protoc flags are pain to add, maintain, integrate and they also make using protoc complicated, so the we're pretty strict about not adding more flags unless they are absolutely necessary. If we introduces a flag for every feature we would have 1000 flags right now.

**Comment by jtattermusch** [comment]
One idea is to add  (if there is really no way to detect whether the compiler supports the `#nullable` define)

```
#if GOOGLE_PROTOBUF_DISABLE_NULLABLE_REFERENCES
// For now Google.Protobuf doesn't have special support for nullable references,
// but this at least allow projects that already use the nullable reference feature from C# 8
// to include protobuf generated code more easily.
#nullable disable
#endif
```

Would that be something that is helpful? (It would require the project to define the GOOGLE_PROTOBUF_DISABLE_NULLABLE_REFERENCES property).

**Comment by jtattermusch** [comment]
CC @JamesNK

**Comment by YairHalberstadt** [comment]
@jtattermusch 

Why not use that same technique, but to add nullable annotations in the correct places, guarded by an ifdef?

It would make the generated code much longer and uglier, but I don't know if that's a concern.

**Comment by jtattermusch** [comment]
> @jtattermusch
> 
> Why not use that same technique, but to add nullable annotations in the correct places, guarded by an ifdef?
> 
> It would make the generated code much longer and uglier, but I don't know if that's a concern.

The difference is that what I'm proposing is a pretty trivial fix that could potentially unlock lot of users. What you're proposing is a significant time investments and it would need to be designed and tested carefully.

Besides that, much longer and uglier code is definitely a concern.

**Comment by mtrtm** [comment]
Did this thread end up dying?  As a .Net developer, I was shocked to learn nullable wasn't supported.  There were some arguments a year ago about c# 8, but we are about to be on 9.  I would love if this could be implemented - nullable is an incredibly positive option to be able to turn on.

**Comment by Falco20019** [comment]
@jtattermusch We are also waiting and hoping for this as it's blocking us from enabling nullable annotations in our code base without getting lot's of errors or enabling nullability in each of our files manually.

**Comment by SonnyX** [comment]
#7382 has seemingly been merged, just waiting for a new release.

**Comment by Falco20019** [comment]
@SonnyX Are you sure that's related? I think that's only about optional-marked fields.

@jtattermusch C#9 is around the corner. Any plans or a roadmap here?

**Comment by Falco20019** [comment]
As of the important box in https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references#nullable-contexts this would need to explicitly set `#nullable enable` since it has the `<auto-generated>` tag in it's file comment. Just wanted to leave this in reference to https://github.com/protocolbuffers/protobuf/issues/6632#issuecomment-617584118 since it will be disabled automatically, even with the `<Nullable>enable</Nullable>` being automatically defined in the `csproj` templates for .NET 6.

**Comment by cwe1ss** [comment]
> As of 3 weeks from now .NET 4.8 and earlier are all unsupported. .NET 5 will also be unsupported and all versions of .NET Core that don't support nullable types are already unsupported.

I'd love to see nullable-support as well, but I don't think your statements regarding End of Support for older .NET versions are correct:
* https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core
  * .NET Core 3.1 has End of Support on December 3, 2022
  * .NET 5 on May 8, 2022
* https://docs.microsoft.com/en-us/lifecycle/products/microsoft-net-framework
  * .NET 3.5 SP1 on Jan 9, 2029
  * .NET 4.6.2+ versions don't have any end dates yet

**Comment by runenilsenoe** [comment]
Still waiting for this

**Comment by ps-weber** [comment]
I think the problem most people are interested in is avoiding assigning null to string fields.

As far as I can see this could simply be done with the `DisallowNullAttribute` behind a preprocessor definition:

```csharp
#if NETCOREAPP3_0_OR_GREATER || NETSTANDARD2_1_OR_GREATER
        [global::System.Diagnostics.CodeAnalysis.DisallowNullAttribute]
#endif
        public string MyStringField
        ...
```

This should be a simple, backwards compatible change that only affects code with nullable analysis enabled.

**Comment by ogxd** [comment]
I've made a PR that implements the solution proposed by @Falco20019, which is in my opinion the most reasonable and effective option (no breaking changes, backward compatible, and optional)
https://github.com/protocolbuffers/protobuf/pull/13218

**Comment by JohnGalt1717** [comment]
Whoot whoot!

**Comment by Machibuse** [comment]
Any news here? it would be great to have this feature!

**Comment by altso** [comment]
Would love to see this feature as well.

**Comment by aomader** [comment]
+1

**Comment by alikleit** [comment]
ping

**Comment by tonydnewell** [comment]
This is still a work in progress - see https://github.com/protocolbuffers/protobuf/pull/16240

**Comment by mwasplund** [comment]
I may be missing some nuance of the original conversation, but I am wondering if this change to enable nullable annotations could also take on allowing for optional properties to be nullable. https://github.com/protocolbuffers/protobuf/issues/9083 If I am reading it correctly, the issues was that changing the behavior of explicit optional fields was a breaking change. Could we use the new enable_nrt to piggyback on for this breaking change as well?

**Comment by jskeet** [comment]
@mwasplund: No, they're *very* different kinds of change. One can enable nullable reference types - in environments that support it - with no runtime changes at all. The types advertised remain the same, just with different annotations. The changes proposed on #9083 would be incompatible with previously-published versions of libraries - whereas enabling NRTs wouldn't.

But more than that, that sort of differentiation simply isn't the direction that protobuf as a whole intends to go. See https://github.com/protocolbuffers/protobuf/issues/9083#issuecomment-2120790708 for more details. The proposal in #9083 is *not* going to happen.

**Comment by mwasplund** [comment]
Both of these proposed changes have compile time incompatibility, however should have effectively the same runtime characteristics.

**Comment by jskeet** [comment]
@mwasplund: I really, really don't want this issue about nullable reference types to end up being more about a feature that I don't think we'll ever implement. Please, please just accept "no" for an answer on this. (I have *very limited* time to work on protobuf at all - the more time I need to spend justifying a decision that has already been made, the less time I'll have to actually work on changes.)

**Comment by mwasplund** [comment]
I do not think it is too much to ask for clarification on your previous decision in light of the new context around this issue when you highlighted breaking changes behind compiler flags being a reason to not do the work.

In my naive outsider perspective this seems like a cheap easy change that would be a nice quality of life improvement for C# developers to use well established nullable optional patterns in our ecosystem instead of littering my code with has checks.  
```csharp
var visualType = input.HasVisualType ? input.VisualType : null;
```

I would have been more than happy to contribute my time to implement this feature and with an open conversation we may have been able to come to a shared design that meets both our needs (third property that lives alongside the other two?). I accept your decision and will not respond anymore. I am sorry to have wasted your time.

**Comment by Falco20019** [comment]
My main issue with the current solution is that there are `HasXxx` for fields, but on repeated and maps there is no proposed way to check. If you have a `repeated google.protobuf.StringValue entries = 1` there is no method to check.

If https://github.com/protocolbuffers/protobuf/pull/16240 is still a valid option for @jskeet (it wasn't closed at least yet), I can try to fix the remaining points.

**Comment by jskeet** [comment]
Repeated and map fields will never be null - you can just check for the count of use Any()

**Comment by Falco20019** [comment]
@jskeet You missed my point. It's not about `entries == null` but `enries[2] == null`.

**Comment by jskeet** [comment]
@Falco20019: Ah, right. IIRC (and I'd need to check in detail) `entries[2]` can never be null in repeated fields or maps. `StringValue` is the one thing I'd check against, but given that `RepeatedField.Add` immediately calls `ProtoPreconditions.CheckNotNullUnconstrained`, I think that it *still* can't be null. (Ditto for `MapField`.) If there's something I've missed, please let me know. (It's been a while since I've been in the weeds on this.)

**Comment by jakoss** [comment]
> I do not think it is too much to ask for clarification on your previous decision in light of the new context around this issue when you highlighted breaking changes behind compiler flags being a reason to not do the work.
> 
> In my naive outsider perspective this seems like a cheap easy change that would be a nice quality of life improvement for C# developers to use well established nullable optional patterns in our ecosystem instead of littering my code with has checks.  
> ```csharp
> var visualType = input.HasVisualType ? input.VisualType : null;
> ```
> 
> I would have been more than happy to contribute my time to implement this feature and with an open conversation we may have been able to come to a shared design that meets both our needs (third property that lives alongside the other two?). I accept your decision and will not respond anymore. I am sorry to have wasted your time.

I do agree. The whole point of nullable annotations is to have compile help us spot missing values. If we mark everything as non nullable then compiler won't help us with nothing. Effectively this feature won't change a single bit for the users, we will still workaround the same issue

**Comment by Falco20019** [comment]
@jskeet Related to this, C# is not behaving like described in https://protobuf.dev/programming-guides/nullable-getters-setters/ which is why I don't fully understand the explanation in https://github.com/protocolbuffers/protobuf/issues/9083#issuecomment-2120790708

Trying `msg.Child.Grandchild.Foo == 0` will throw an exception and there's no difference in the properties and backing fields being generated for implicit or explicit field presence in C# for messages. So I do understand why everyone is confused.

Hazzers are also never generated for message-typed fields, so checking them for `null` is the correct way.

I used the following since with the `[default = 72]` I will get `Explicit default values are not allowed in proto3.` when using `Grpc.Tools 2.69.0` (latest):
```proto
syntax = "proto3";
package test;

message Msg { optional Child child = 1; }
message Child { optional Grandchild grandchild = 1; }
message Grandchild { optional int32 foo = 1; }
```

I also checked it with `proto2` which allowed `default` option and it also made not difference. Same for `edition = "2023"`. So maybe fixing explicit field presence would be the alternative to adding nullable in consideration of the given design choice?

**Comment by jskeet** [comment]
Sorry, I *really, really* don't have time to go into this. I'll ask someone to have a look at that documentation (as it really shouldn't mention C# - it's probably accurate for Kotlin and Rust).

To explain my position: I'm not in the protobuf team. I maintain the .NET protobuf runtime as best I can with very little time available, prioritizing the needs of the Google Cloud client libraries. I understand that it's frustrating, but my alternatives to saying "Sorry, I haven't any time" are either to ignore these issues entirely, or not do the job that I'm actually meant to be doing.

**Comment by Falco20019** [comment]
Sorry if you took it personal. I just didn't know who else to mention. If there's someone else to figure this out, feel free to let me know and to ignore my follow-up messages (as it might even be better to have this on a separate issue). Just giving some more context on this.

https://protobuf.dev/programming-guides/field_presence/#semantic-differences mentions that for explicit presence, there should be `HasXxx` and `ClearXxx` methods. https://protobuf.dev/programming-guides/field_presence/#presence-in-proto3-apis mentions that messages are always be explicit presence (independant of `optional` being added). But C# is only providing a nullable property that's not declared `null`.

**Comment by jakoss** [comment]
I think nobody should take it personal, I'm pretty sure we want to achieve the same thing. I'm also sure that Jon doesn't have to work on this himself (i feel like we have many potential contributors here), but we need the green light from some maintainer to achieve better protobuf experience for .NET devs

**Comment by Falco20019** [comment]
I also already did offer to help with https://github.com/protocolbuffers/protobuf/pull/16240 3 weeks ago. This offer is still up.

https://protobuf.dev/programming-guides/field_presence/#c-example-1 has the example for C# and in my opinion having the hazzers makes totally sense. So I'm not even fully on the same page with most here to make numerics or enums using NRTs. I think they work as designed and that's good. `google.protobuf.StringValue` is a separate case and having that generate as `string?` would be nice as it's well-known, but I'm even fine with it as-is.

My real pain point is just how messages are implemented. Even though they are using explicit presence and are implemented as nullables, they are not showing as nullable. So we over and over run into people accessing them (last time a bug appeared today in our code base due to this) without checking for `null` as the compiler is fine with it. So you have to explicitly KNOW that these don't have hazzers but you will need them checked for `null`.

**Comment by jakoss** [comment]
I agree, for numerics and enums NRTs are not that important (REST apis and standard JSON serialization libraries have the same issue, proto is not unique in this way). But like you said - StringValue not being nullable seems silly. Also we have wrappers for numerics like `google.protobuf.Int32` which in my opinion should be nullable as well. I feel like this way the whole idea being wrapper types in proto.

Fully agree about messages as well. And i think if we do introduce nullable annotations but do not change this behavior - it will get even more confusing. Because at this point users will have compiler warning stating they should not null check variable since it's marked as non-nullable (but no compiler warning they should call HasXxx instead...)

**Comment by Falco20019** [comment]
To not trigger Jon any more, I tried to keep https://github.com/protocolbuffers/protobuf/pull/20242 addressed to the team in general and not him directly. I'm sorry that he was added anyway due to the CODEOWNER file.

The PR is based on what he is intending to look into whenever he finds time. I will be using it internally for now until there's an official fix, so I will stop bugging everyone here. If someone is interested, feel free to use the branch and compile it. Use `csharp_opt=enable_nrt` to enable the feature. It's fixing `google.protobuf.StringValue` to `string?` and mark messages as nullable :)

**Comment by jskeet** [comment]
@Falco20019: I appreciate that, thanks. I'll probably need to find personal time to review that in, but I *will* try to get to it at some point. I have one immediate comment, but everything else will wait.

**Comment by JohnGalt1717** [comment]
The real solution to this is just to add nullable on top of optional for a new major version of protobuf.

Optional should require a default value or nullable = true.

Languages that don't support nullable types will have to deal with nulls in their values as nullable won't do anything in their language generation, but then that's what they have to do now.  Optional with default value all languages can support just fine.

This should be super high priority for next version of protobuf because since Protobuf was originally created almost every typed language now has nullable types (because they're a no brainer) so this is a major impediment to adoption outside of microservices and even then, it's a major issue that results in things breaking that shouldn't be breaking and would be caught at compile time.

Still don't understand the pushback for this basic functionality to be supported in a modern serialization system.

**Comment by mkruskal-google** [comment]
@Falco20019 thanks for taking over that PR, and pointing out the confusion in our documentation.  I will update our docs to call out the non-standard behavior of C# message fields being nullable.  Since message fields can't have defaults anyway, it's mostly an ergonomic issue (in the context of that doc).  We could leverage editions to change them to always be non-null like other languages in a non-breaking way (@jskeet), but adding an option for nullable annotations seems like a simpler solution for now

**Comment by esrauchg** [comment]
@JohnGalt1717 Just want to mention that https://protobuf.dev/programming-guides/nullable-getters-setters/ was written after your previous round of feedback on this topic. It unfortunately did not properly clarify the state of C# message-field accessors properly which we will follow up on, but otherwise intends to address your comments.

One important thing to stress is that Edition 2023 removes the optional keyword and all (non-repeated non-map) fields simply have presence by default, with a feature flag to opt out back to proto3 presence behavior (allowing for semantics preserving upgrades from proto3 syntax to Edition 2023). It also restores the ability to choose nonzero default values for primitive fields.

Your worldview is reasonably influenced by your personal experiences using C# Proto with Proto3, where there aren't immutable defaultinstances of messages, and where most int fields do not have presence, and schema owners would need to take some affirmative decision to mark something optional. It is a world where it wouldn't be odd if one file has some int32 fields that are optional and some that are not (and, someone actively chose it to be that way if it is). If that was how everyone used Protobuf, it very well might be sensible first-class optional/nullable behavior on int accessors in languages like Kotlin, Rust and C#.

Under the usage patterns that we expect to be most common looking forward (which are already very common for some folks, including anyone who stuck with Proto2), the world is just quite different than that. By having all int accessors just return a simple int (which may be a nonzero default), it delivers on the code ergonomics that effectively match proto3 implicit-presence ergonomics of being able to use fields and not have to think about presence if they don't want to, and it also more smoothly enable folks to move from their Proto3-default behavior to Editions-default behavior without suddenly adding null handling on every field.

We appreciate the feedback that vision doesn't line up with your own usage patterns of Protobuf, but this design direction is something that experienced engineers have arrived on after carefully looking at how Protobuf is used inside and outside of Google, including what would be the lowest impediment to adoption outside of microservices. It is not a no-brainer that direction is wron, and we're realistically not going to take a drastically different direction in any immediately upcoming major version change.

Thanks!

**Comment by JohnGalt1717** [comment]
You're still not addressing the obvious: nullable types across all languages that support them and IsNull or something else for the rare few backwards languages that don't.

You're making proto unsafe compared to what it could be and thus not used as a result.  I would never use optional the way it was designed, but it doesn't address nullability and expressing it intelligently in modern languages.

Null is a real, valuable thing. Ignoring is is harming protobuf and thus grpc adoption and trying to punt it as not required is just a waste of everyone's time.  Even all of Google's endless languages have null and have nullable types at this point, so being able to express null in the serialization language and thus enable every language to use their nullable types is a no brainer.

It's this kind of stubbornness that is the undoing of sooooo many Google projects that languish and die.

**Comment by Falco20019** [comment]
You're still missing the point. If you want to have only "real nullables", using optional isn't the correct way. Just go and use the wrapper types everywhere. You will get exactly what you want and it will work across all languages. The outer message basically gives you the presence part (null or not) while the inner value (implicit presence) gives you the current value that can't distinguish if it was set or always on this possibly default value.

The whole point of optional (and explicit presence) is that you can easily access the current value (which is if unset the defined default value and not always null) but still be able to see if the value was set or not. An optional int32 therefore is really an int that was changed or keeps it default (which could change over time) and not a nullable int that has a fixed value with presence flag.

Also the point is that it makes semantically a difference. I'm not working for Google or Microsoft and DO use this feature as intended to be able to detect on the receiving end of a field was touched or not. With explicit presence you can easily achieve scenarios like merging messages by only transmitting the changed object. With implicit presence or just nullables you can't tell anymore as the presence of non-nullables is lost in transmission. You would need to explicitly transmit a field mask as well on what was changed.

For some it's a nuisance, for others it's the reason they chose ProtocolBuffers over other protocols like MsgPack.

Don't get me wrong, I still find it annoying that nullability annotations are missing at the moment as it also leads to bugs in our software. You can't rely on compiler warnings to get all nullability issues right now. But that's the fixable part that's addressed. Adding annotations that are missing, not dropping a key feature.

**Comment by esrauchg** [comment]
As this Issue is about C# nullable annotations, lets keep any further discussion on this thread on that topic (we are very interested in finding the path to solve this particular pain point).

The issue about nullable int accessors was discussed at length in https://github.com/protocolbuffers/protobuf/issues/9083 and it was closed as not planned; significant new evidence is unfortunately not going to reopen the topic; lets try to avoid rehashing the same arguments that have already been considered.

Thanks!

**Comment by ngbrown** [comment]
Keep open until pull request is merged and this is resolved.

**Comment by Karlosbubi** [comment]
since #9083 is closed, where is the current state of nullability documented ??
Like whether or not the client will throw if i assign `null` to an optional field.

**Comment by Falco20019** [comment]
#20242 is the current PR to get nullability in, but sadly no progress since June (although PR is reviewed and ready to be merged by the team).

**Comment by esrauchg** [comment]
(Pasting same update here as put on the PR)

Sorry about slow updates here. This is still actively (slowly) being discussed within Google.

I can give an update about why this is not straightforward and where we are here.

The main topic is regarding .NET Framework users who cannot compile the code which has nullable annotations in it. Broadly Protobuf's standard stance is that we set our minimum supported language/compiler/platform/whatever, and we constrain our feature set to that version. For example in Java we support Java 8 minimum, we don't do anything which is not Java 8 compatible. Inside our runtimes we can sometimes have conditional branches for newer runtimes to get better behavior in new environment while maintaining our compatibility, with only one set of source files.

Unfortunately this is not really possible with C# nullable annotations: there is no way we can have the gencode have those annotations while still having it be source compatible with our complete support matrix.

The second topic is, given this is sufficiently important, how can we give people a lever to opt into "not compatible with our complete support matrix". Since doing so is unprecedented across all languages, we don't have a good playbook of precedents to follow for how to do it.

The PR proposes a protoc flag; this functionally works, but protoc flags are something that the Protobuf team would like to avoid essentially at all costs. It has many problems, but one major one is that the protoc design is that people are expected to run protoc many times in isolation (each time with a -I for its includes). Each invocation will create gencode which references other gencode produced independently by a different invocation. The gencode basically has to just 'know' what the other invocation is going to produce for the final gencode to successfully compile, and any flag that influences gencode creates compatibility problems based on the combination. And its an extremely difficult property to test because you get into combinatorial support matrix problems of "one set of gencode was made with protoc 26 with flags [x, y, z] enabled and has outgoing references to other imports, the imports are created with protoc 33 which was run with flags [a, b, c] enabled, does that work or not?".

Two other alternatives that have been discussed internally are:

Options in .proto files. These days it would probably mean an Edition Feature, where each .proto file could then opt in to the new behavior and the corresponding gencode only works on correspondingly supported versions.

#ifdefs PROTOBUF_ENABLE_NRT in the gencode which you can turn on in the csproj, effectively making it so the gencode itself is the same no matter what and then the entire project is guaranteed to have a consistent state of it all on-or-off.

Broadly speaking the current discussion seems to be leaning towards that this should be supported as an option in the .proto file as a better solution compared to a protoc-flag or #ifdefs, but there's not clear consensus yet.

We are open to more feedback about those or other possible options here to help inform next steps.

Thanks!

**Comment by JamesFieldist** [comment]
Have 2 code paths for .net:  .net Framwork (4.8 etc) and .net

The .net compiler would be nullable.  
The .netFramework compiler wouldn't be.

They are 2 different targets.  All versions of .net that don't support nullables are out of support from Microsoft so there's no support case where this is an issue. And for the old stuff, just keep the compiler the same as it is.

Bonus points for using record not class for .net and making them fully immutable.

**Comment by Falco20019** [comment]
That's basically what was intended through Grpc.Tools based on the target framework (only setting the flag for .NET). But I do understand the concerns made.

Using records and immutability would basically mean completely redoing the generator. So probably not happening.

**Comment by luryus** [comment]
I built (well, LLMs mostly did) an MSBuild thing that allows playing around with these changes while we continue to wait for the upstream protoc implementation: https://github.com/luryus/ProtobufNrtAnnotator/

This is an MSbuild task that hooks into the build process after Grpc.Tools and injects nullable annotations directly into the gencode. Dumb and fragile, ...but it kinda works?

I enabled this for our company-internal protos (a few dozen gRPC services). I haven't really tested this much yet, but one immediate outcome is that there's _a lot_ of new warning noise from message fields ("Dereference of a possibly null reference"). I of course knew this would be the case, but just the amount of warnings in at least our code caught me off guard. For example, take this simple gRPC scenario:

```protobuf
message Foo {
  int32 id = 1;
}
message FooRequest {}
message FooResponse {
  Foo foo = 1;
}

service FooProvider {
  rpc GetFoo(FooRequest) returns (FooResponse) {}
}
```

Now every single client that calls GetFoo will see a new warning from accessing `response.Foo.Id`, if they did not already verify that Foo is not null. I know this is technically correct, but I'm not so sure this is what we would actually want in practice. At least with our internal gRPC services, where only one implementation of the service exists, we can assume that certain message fields are always present (e.g. here the FooProvider implementation would always populate the `foo` field in its responses). And therefore our clients currently do not necessarily have null checks in place for those fields.

I worry that if the NRT feature is shipped in something like the current state (a flag that affects all fields and messages), the need for an escape hatch to control the annotations for individual fields will crop up very soon. Depends on the use cases and usage patterns of course. This is just my experience based on one codebase, not necessarily representative of all C# protobuf code everywhere.

> 1. Options in .proto files. These days it would probably mean an Edition Feature, where each .proto file could then opt in to the new behavior and the corresponding gencode only works on correspondingly supported versions.

If an edition feature was implemented, could it be supported on fields or messages as well? That would allow certain fields or messages to stay null-oblivious. While not a perfect solution, it could at least help reduce the warning noise in some situations. And the most important nullability annotations would still be present (StringValue/string is the most obvious one).

One possibility would be to allow overriding the nullability of individual fields in the proto definition. A csharp-specific field option could work, but I don't know if those are a thing?

**Comment by esrauchg** [comment]
Thank you for the useful feedback! It is always great to get this exact shape of feedback before a change is made instead of only after.

> If an edition feature was implemented, could it be supported on fields or messages as well? That would allow certain fields or messages to stay null-oblivious. 

With an edition feature we could do this (a file-level setting and then also field-granularity; thats how many Editions features work today), and it seems like might be a reason to do it that way if that is desirable behavior.

We still have an active (slow/careful) discussion going internally about the tradeoffs here; I think we are headed towards Edition feature for this. The main concern is if we only added it with Edition 2026 that it would be annoying for people to adopt today (Editions features are not available in Proto3 syntax). We still intend to publish a tool called prototiller which would rewrite a Proto3 file to an Editions file in a perfectly-semantics-preserving way which would make that 'easy' (run the tool with no impact on gencode shape or runtime behavior, and then from there you can add options) but thats still in development as well.

**Comment by Machibuse** [comment]
@luryus 

I wrote something like your solution some month ago. 

https://github.com/Machibuse/Porticle.Grpc.TypeMapper

This is also a build step running after protoc and CAN change the generated files, but only if you give specific comments on top of the Properties.

it can change the generated files like this:

- Protobuf string to C# Guid conversion on the fly
- Protobuf google.Protobuf.StringValue to C# Guid? on the fly
- Protobuf google.Protobuf.StringValue to C# string? 
- Protobuf optional enum to C# nullable enum
- wrap all not nullable proto stings in #nullable enable/disable
- wrap all StringValue fields in #nullable enable/disable and change string to string?

It is no perfect solution but it helps a lot until the grpc c# developers wake up

Feel free to add wishes or help on the project

**Comment by loudar** [comment]
> We still intend to publish a tool called prototiller which would rewrite a Proto3 file to an Editions file in a perfectly-semantics-preserving way which would make that 'easy'

Do you have a timeline or estimated release planned for when Prototiller should be available? It has been mentioned for 4 years now and it's also mentioned a lot in the docs, but there's no other info besides "it can do that" and no tool being available.

**Comment by jskeet** [comment]
Just as a heads-up: work is genuinely happening on this, initially adding an empty c_sharp_features.proto file (with all the required plumbing, which is non-trivial) and then putting the feature itself in.

**Comment by jskeet** [comment]
@esrauchg and @luryus 

> With an edition feature we could do this (a file-level setting and then also field-granularity; thats how many Editions features work today), and it seems like might be a reason to do it that way if that is desirable behavior.

I'm going to investigate this. There are three immediate implications of doing it at the field granularity:

1. We'd end up enabling annotations for each enabled property:

```csharp
#nullable enable annotations
<property code>
#nullable restore annotations
```

This is ugly for anyone looking at the generated source code, but not disastrous. I don't *think* it's then a problem that the rest of the code (e.g. merging) wouldn't have annotations enabled.

2. As a change from the current PR, we wouldn't end up with the `?` in things like `override bool Equals(object other)`.

Again, I don't think that's a problem, but it's a difference.

3. Without annotations enabled, the copy constructor `MyMessage(MyMessage)` wouldn't catch problems where null was passed in incorrectly.

I suspect we can probably address this by observing this at both message and field granularity - where setting the feature for the file would enable it for all messages; setting it for a message would enable it for all fields *and message-level methods* (removing the difference for 2 and 3). Setting it for just specific fields would result in just the per-property annotations. Setting it for the message but *disabling* it for specific fields may be fiddly/weird... I might just say that's not supported to start with, if it's too hard to get right immediately.

I think that's probably going to be the best solution, but requires a bit more time than just using the existing PR. @esrauchg and @mkruskal-google I may well need to consult you...

**Comment by jskeet** [comment]
This has been implemented (at a .proto level granularity) for edition 2026 in #27762.

**Comment by Falco20019** [comment]
@jskeet Thanks for the implementation! The major downside for us with using editions is that we also have to support Rust which (as of today) does not even support the `editions` syntax at all.

@esrauchg just yesterday pushed `google-protobuf@0.36.2` / `protobuf@4.36.2` which is still in beta and why our Rust team is still reluctant switching. I assume that this will be up2date and support editions as it's a full rewrite. Is there a roadmap on when this is expected to be production-ready? And will it support 2026 edition around the same time C# will offer it? Before the 2026 is out, it's not an issue for us anyway, we just would love to know if we can plan using it when available 👍

**Comment by esrauchg** [comment]
> @esrauchg just yesterday pushed google-protobuf@0.36.2 / protobuf@4.36.2 which is still in beta and why our Rust team is still reluctant switching. I assume that this will be up2date and support editions as it's a full rewrite. Is there a roadmap on when this is expected to be production-ready? And will it support 2026 edition around the same time C# will offer it? Before the 2026 is out, it's not an issue for us anyway, we just would love to know if we can plan using it when available 

We'll have some comms but with 37 (our next minor release) we will be declaring google-protobuf Rust to be 'production-ready' despite it being `0.37` versioned. The `0.` is based on that we _might_ do corner case api breaking changes within the next two years and don't want to burn major versions so fast (basically, any breaks we do should not be breaks for almost anyone in practice but we want to take strict semver seriously; we expect to be more stable than Prost as the baseline, which is otherwise living on 0.x).

Can you clarify if your Rust usage is currently on Prost or on `protobuf` crate on the 3.x line? With appologies that the overall state of ecosystem with RustProto is all a bit awkward unfortunately (Prost is only reactively maintained, `protobuf` crate 3.x is terminal, and our new API is a bit strange since we are prioritizing zero-copy C++ interop behind the api)

**Comment by Falco20019** [comment]
> Can you clarify if your Rust usage is currently on Prost or on `protobuf` crate on the 3.x line? With appologies that the overall state of ecosystem with RustProto is all a bit awkward unfortunately (Prost is only reactively maintained, `protobuf` crate 3.x is terminal, and our new API is a bit strange since we are prioritizing zero-copy C++ interop behind the api)

Right now we live on protobuf 3.x and look for a migration target as it's, as you mentioned, going to be end of life. In their repository they mentions that there will be an official as 4.x which shows to be basically your 0.x
